bottlecap/src/telemetry/listener.rs

```diff
@@ -18,21 +18,62 @@ pub struct HttpRequestParser {
 
 const CR: u8 = b'\r';
 const LR: u8 = b'\n';
+/// It is guaranteed that the headers will be less than 256 bytes.
+const HEADERS_BUFFER_SIZE: usize = 256;
 
 impl HttpRequestParser {
-    /// Create a `HttpRequestParser` from passed `buf`
+    /// Create a `HttpRequestParser` from a `TcpStream`
     ///
     /// # Errors
     ///
-    /// Function will fail if parsing of headers or body in `buf` fail.
-    pub fn from_buf(buf: &[u8]) -> Result<HttpRequestParser, Box<dyn Error>> {
+    /// Function will error if the stream cannot be read from.
+    ///
+    /// It will also error if the headers cannot be parsed.
+    ///
+    /// Or if the body cannot be parsed.
+    pub fn from_stream(mut stream: &TcpStream) -> Result<HttpRequestParser, Box<dyn Error>> {
+        stream.set_nonblocking(true)?;
         let mut parser = HttpRequestParser {
             headers: HashMap::new(),
             body: String::new(),
         };
 
-        let body_start_index = parser.parse_headers(buf)?;
-        parser.parse_body(buf, body_start_index)?;
+        let mut headers_buf = [0u8; HEADERS_BUFFER_SIZE];
+        loop {
+            match stream.read(&mut headers_buf) {
+                Ok(_) => {}
+                Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                    continue;
+                }
+                Err(e) => {
+                    error!("Error reading from stream: {}", e);
+                }
+            }
+
+            let _ = parser.parse_headers(&headers_buf);
+            if parser.headers.contains_key("content-length") {
+                break;
+            }
+        }
+
+        let body_start_index = parser.parse_headers(&headers_buf)?;
+        let content_length = parser
+            .headers
+            .get("content-length")
+            .expect("unfallible")
```

Comment:
do we want to expect here? can't we just drop the request if it's malformed or something broke in our assumption?

Reply:
The loop before that ensures that the key exist, which means that this is infallible, wdyt?

Reply:
hmm, I think that's okay. It just makes me nervous

Reply:
I could definitely add another `if let Some(...)` block if that makes you feel more confident